src/test/java/uk/gov/pay/products/service/ProductApiTokenManagerTest.java

```diff
@@ -0,0 +1,132 @@
+        setUpPublicAuthToReturnResponse(NEW_API_TOKEN);
+
+        String apiToken = productApiTokenManager.getNewApiTokenFromPublicAuth(demoPayment);
+
+        assertThat(apiToken, is(NEW_API_TOKEN));
+    }
+
+    @Test
+    public void shouldThrowExceptionWhenPublicAuthFailsToGenerateAnApiToken() {
+        setUpPublicAuthToReturnResponse(null);
+
+        MissingApiTokenException e = assertThrows(MissingApiTokenException.class,
+                () -> productApiTokenManager.getNewApiTokenFromPublicAuth(demoPayment));
+
+        assertThat(e.getMessage(),
+                is(String.format("Failed to get a new API token for product %s of type %s", demoPayment.getExternalId(), demoPayment.getType())));
+    }
+
+    @Test
+    public void shouldRotateApiTokenForAProductWhichExists() {
+        Product modifiedDemoPayment = ProductEntityFixture.aProductEntity().withApiToken(NEW_API_TOKEN).build().toProduct();
+        ProductFinder productFinder = mock(ProductFinder.class);
+        when(productFactory.productFinder()).thenReturn(productFinder);
+        when(productFinder.updatePayApiTokenByExternalId(demoPayment.getExternalId(), NEW_API_TOKEN)).thenReturn(Optional.of(modifiedDemoPayment));
+
+        productApiTokenManager.rotateApiTokenForAProduct(demoPayment, NEW_API_TOKEN);
+
+        verify(mockAppender, times(1)).doAppend(loggingEventArgumentCaptor.capture());
+        List<LoggingEvent> loggingEvents = loggingEventArgumentCaptor.getAllValues();
+        assertThat(loggingEvents.get(0).getFormattedMessage(),
+                is(String.format("Regenerated API token for product %s of type %s", modifiedDemoPayment.getExternalId(), modifiedDemoPayment.getType())));
+    }
+
+    @Test
+    public void shouldThrowExceptionWhenRotatingApiTokenForAProductWhichDoesNotExist() {
+        ProductFinder productFinder = mock(ProductFinder.class);
+        when(productFactory.productFinder()).thenReturn(productFinder);
+        when(productFinder.updatePayApiTokenByExternalId(demoPayment.getExternalId(), NEW_API_TOKEN)).thenReturn(Optional.empty());
+
+        FailToRotateApiTokenException e = assertThrows(FailToRotateApiTokenException.class,
+                () -> productApiTokenManager.rotateApiTokenForAProduct(demoPayment, NEW_API_TOKEN));
+
+        assertThat(e.getMessage(),
+                is(String.format("Failed to rotate API token for product %s of type %s", demoPayment.getExternalId(), demoPayment.getType())));
+    }
+
+    private void setUpPublicAuthToReturnResponse(String expectedNewApiToken) {
+        WebTarget webTarget = mock(WebTarget.class);
+        when(client.target(PUBLICAUTH_URL)).thenReturn(webTarget);
+        Invocation.Builder builder = mock(Invocation.Builder.class);
+        when(webTarget.request()).thenReturn(builder);
+        Response publicAuthResponse = mock(Response.class);
+        when(builder.post(any(Entity.class))).thenReturn(publicAuthResponse);
```

Comment:
Can we make this a bit more rigorous? It would be good to check we’re sending the right payload (including the description) for each of the three types of products. We could have three variants of `shouldReturnANewApiTokenWhenPublicAuthGeneratesTheToken` to facilitate this.

Reply:
Yes, I have made this a bit more rigorous.